### PR TITLE
Ancestral Calendar

### PR DESCRIPTION
**Pull request by harrislineage**
This is the initial draft for an ancestral calendar of events. I am currently processing 10 generations of ancestors for birth and death events and sorting them into collapsible containers by month.

Priority:
- Sort events within months (2 March 1752 should come before 2 March 1804)
- Style content containers (mimic profile previews?)

Other:
- Add relationship information (how am I connected to this person?)
- Expand/Collapse all container button(s)?
- Collapse other containers when a new container is opened?
- Auto-open and focus on this month's container?

image

### CODE REVIEW

**Comment by bcaseyrls** [comment]
Nice work. I'd been considering a calendar view (traditional grid). Maybe that would be an optional display mode here? I imagine much could be done with just CSS. You might also consider optionally loading a person's watchlist instead of 10 ten generations of ancestors. Maybe a way to highlight people that are still living? There's a feature coming soon to the API that will let you load nuclear relatives (parents, siblings, spouses, children) out to N generations that may be useful here.

**Comment by harrislineage** [comment]
Good feedback @bcaseyrls. Let me get to a good stopping point with the initial view I have, and then I can branch out to other formats/data sets.

**Comment by bcaseyrls** [comment]
This looks great. I think the switch to a "pane" view with the list of month names at the top is much easier to navigate than the open/close containers from before. Nicely done. A few comments.

As a minor thing, I'd consider making the list of months at the top bigger or bolder or something. Or I'd style them more fully as links with colors. That definitely doesn't have to be done immediately, it's just a comment.

Something I think should be addressed, though possibly not here, is that the page-top headline isn't accurate for this view. It reads "Family Anniversaries for ...." and uses the name and WikiTree ID from the form. But it's always displaying the watchlist/family of the logged in user. Inputting a different WikiTree ID doesn't change anything (and shouldn't - you can't read watchlists for other profiles), but it makes the headline incorrect and confusing. 

This is probably something we need to fix in index.html and tree.js since it's common code modifying that headline (using the View::meta() stuff in ViewRegistry::initView()). If you see an easy workaround through the calendar view to address the headline, we can do that. Otherwise, I'll merge this in and we can create a separate issue to address the needs of an "atypical" view headline.

File: views/calendar/calendar.js [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                        &nbsp;<span style="text-decoration:underline;cursor:pointer;" id="month-Oct">October</span>&nbsp;
+                        &nbsp;<span style="text-decoration:underline;cursor:pointer;" id="month-Nov">November</span>&nbsp;
+                        &nbsp;<span style="text-decoration:underline;cursor:pointer;" id="month-Dec">December</span>&nbsp;
+                    </div><br>
+                    <div class="loader">
+                        <image width="200px" src="views/calendar/wtloadericon.png" />
+                    </div>
+                    <div class="calendar">
+                        <div id="Jan" class="content"><h2>January</h2></div>
+                        <div id="Feb" class="content"><h2>February</h2></div>
```
Comment (MichalVasut) [inline views/calendar/calendar.js:58]: I've noticed, there are several places using month names (either strings or objects)

Maybe better way is to define constant with array of those names and then just use it as a base in combination with map and join functions.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join

You can reduce those names to 3 letters using map and slice functions

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice

As for array to object, check this

https://stackoverflow.com/a/36388401

File: views/calendar/calendar.js [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+                        }
+                    });
+                }
+                // Then we insert the content into the correct month
+                // This could be simplified...
+                for (var i = 0; i < x.length; i++) {
+                    try {
+                        $(`#Jan`).append(`
+                        <span>${AllMonths.Jan[i].day} ${getMonth(AllMonths.Jan[i].date)}
+                        ${AllMonths.Jan[i].year} <a href="https://www.wikitree.com/wiki/${AllMonths.Jan[i].id}">
```
Comment (MichalVasut) [inline views/calendar/calendar.js:139]: Those looks the same, here would the map function on array also work, just apply own function, that receives month name as parameter and produces that span. (or appends it directly, without returning)

File: views/calendar/calendar.js [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            $("span[id^='month-']").click(function (e) {
+                $(".calendar").children().removeClass("active box rounded orange").css("display", "none").hide();
+                $(`#${e.target.id.replace("month-", "")}`)
+                    .addClass("active box rounded orange")
+                    .css("display", "block");
+            });
+
+            function pushEvents(person, date, text, link, spouse) {
+                if (getMonth(date) == "Jan" && getMonth(date) != undefined) {
+                    AllMonths.Jan.push({
```
Comment (MichalVasut) [inline views/calendar/calendar.js:281]: No need to do it separately for every month, object properties can be accessed using brackets and keys, not only using dot method.

File: views/calendar/calendar.js [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+                        </span><br>
+                    `);
+                    } catch {}
+                }
+            }
+            // And here is just a bunch of random helpers.
+            // This could use some cleanup and simplification.
+            var months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
+            function getMonth(date) {
+                if (date.split("-")[0] != "0000" && date.split("-")[1] != "00" && date.split("-")[2] != "00") {
```
Comment (MichalVasut) [inline views/calendar/calendar.js:265]: not sure why are you checking complete date, not only month

```js
const [day, month, year] = date.split("-").map(item => int(item))
return month ? month : undefined
```

**Comment by harrislineage** [comment]
Thanks for the review @MichalVasut. I would argue that these are non-issues for a first iteration. You will see where I noted in the code that cleanup and simplification was needed. Much like the FanChart, this view is intended to go through an iterative development process while I work out any kinks and work to simplify and strengthen the code.

If @bcaseyrls agrees, we can move forward with the merge and I can create an issue for the outstanding comments to be worked through in the next iteration.

**Pull request merged**